### ISSUE TITLE
New CLI: Replace argparse with Click

### ISSUE THREAD

**Issue by dchristidis** [rucio/rucio#7253]
### Description

I feel that it’s impossible to achieve a good experience while still working within the limitations of the `argparse` library.

This morning I had a quick look into [Click](https://click.palletsprojects.com/en/stable/) and I’m very impressed with the results. The help output is much closer to what I had imagine and it does so with significantly less boilerplate.

```python
#!/usr/bin/env python3

import click

@click.group()
@click.option('--account')
@click.pass_context
def main(ctx, account):
    ctx.ensure_object(dict)
    ctx.obj['account'] = account
    
@main.group()
def rule():
    pass
@rule.command()
def add():
    pass
@rule.command(name='list')
@click.option('--account')
@click.pass_context
def list_(ctx, account):
    print(f'Root --account: {ctx.obj["account"]}',
          f'Subcommand --account: {account}',
          sep='\n')
@rule.command()
def remove():
    pass
@rule.command()
def show():
    pass
@rule.command()
def update():
    pass

@main.command()
def whoami():
    pass
    
if __name__ == '__main__':
    main()
```

Example usage:
```console
$ ./click --help
Usage: click [OPTIONS] COMMAND [ARGS]...

Options:
  --account TEXT
  --help          Show this message and exit.

Commands:
  rule
  whoami
$ ./click rule --help
Usage: click rule [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  add
  list
  remove
  show
  update
$ ./click rule list --help
Usage: click rule list [OPTIONS]

Options:
  --account TEXT
  --help          Show this message and exit.
$ ./click --account=foo rule list --account=bar
Root --account: foo
Subcommand --account: bar
```

**Comment by dchristidis** [comment]
It also allows tying a command-line option to an environmental variable (e.g. with `@click.option('--account', envvar='RUCIO_ACCOUNT')`):

```console
$ RUCIO_ACCOUNT=foo ./click rule list
Global --account: foo
Local --account: None
```

Maybe we can finally remove this questionable implementation:
https://github.com/rucio/rucio/blob/28eb9bfe9b0b882b214c3a737f92738a28ab1450/lib/rucio/client/commands/command.py#L231-L232

**Comment by dchristidis** [comment]
It also offers superior wrapping of the content, compared to `argparse.RawDescriptionHelpFormatter`.

**Comment by voetberg** [comment]
I absolutely agree. It's pretty clearly the superior choice for python CLI. I would like to see it used for future versions.

**Comment by voetberg** [comment]
To note other alternatives: 
[Python-Fire](https://github.com/google/python-fire) and [Plac](https://github.com/ialbert/plac) which produce a more argparse like help menu but are Very easy implementation wise

**Comment by voetberg** [comment]
I have made a gist comparing a few different options, feel free to either use it as reference of comment on it directly. 

https://gist.github.com/voetberg/528794ddd12b1b1e0dd63b4a165f1d21

**Comment by dchristidis** [comment]
> You can't add descriptions for positional args. That's a pretty big one. Might be worth 2 bullet points.

The [documentation](https://click.palletsprojects.com/en/stable/documentation/#documenting-arguments) offers a rationale and a recommendation. If find it reasonable, to the point that I wouldn’t consider this a con.

We’re already making an effort to restrict positional arguments to things that are obvious (e.g. a replication rule ID, a DID, an RSE, an account). Documenting them is superfluous, even difficult to express in a way that adds value. For example:

```console
$ rucio account show --help
usage: rucio account show [-h] -a ACCOUNT

options:
  -h, --help            show this help message and exit
  -a, --account ACCOUNT
                        Account name
```

**Comment by mlassnig** [comment]
With such things I tend to go by my gut feeling (I know, not very scientific!):

I like this `click` thing for reasons I cannot express coherently, and I like `fire` because it doesn't trigger my OCD. `plac` ... I find a bit weird.

**Comment by voetberg** [comment]
> The [documentation](https://click.palletsprojects.com/en/stable/documentation/#documenting-arguments) offers a rationale and a recommendation. If find it reasonable, to the point that I wouldn’t consider this a con.

I think it's reasonable, I just disagree with the conclusion that anything self-explanatory shouldn't have the option for additional information. You can forgo adding documentation for _any_ argument, I just don't like that's you're de-facto locked out of the possibility for positionals. A matter of personal taste though. 

> I like this click thing for reasons I cannot express coherently, and I like fire because it doesn't trigger my OCD. plac ... I find a bit weird.

This is generally my feeling too. The nesting structure `fire` has is very nice,  pythonic and plays really well with OOP structures. `click` does the same but more smoothly for functional setups over classes. `plac` is just _right out_. Not sure what we should use as a tie-breaker though.

**Comment by bari12** [comment]
I come to the same conclusions. From a package support / usage point of view I also think they are pretty equal.

**Comment by voetberg** [comment]
In interest of making a decision and moving forward, I'd put my weight behind `click` for what it's worth.

**Comment by dchristidis** [comment]
I think the comparison should take three aspects into consideration: the user experience, the developer experience, and the features. Sadly, the first two are inherently subjective. As such, here is my own biased assessment. :laughing:

##### User experience

My impression is that Click can present the same amount of information in fewer lines. Fire resembles man pages, which feels excessive for subcommands.

##### Developer experience

I like Click’s approach (decorators) better than Fire’s. I like Click’s documentation better than Fire’s. And I like that Click’s source code has type hints and looks modern and well-maintained.

##### Features

To be frank, I didn’t investigate long enough to make a fair comparison. One thing I already mentioned: Click can populate keyword arguments from environmental variables. Moreover, Click has a few [additional functions](https://click.palletsprojects.com/en/stable/utils/) beyond CLI structure and help pages, such opening URLs (this will be very helpful for OAuth flows).

**Comment by voetberg** [comment]
I agree with this assessment (with minor disagreements on the dev experience, I found fire's a lot more clean, but this is personal preference.) 

What really I think puts me off fire is that it seems to mostly be a one person project that we shouldn't expect to get extensive feature updates, so if we have to put trust in a non-python core dependency I have more faith in `click` having support in the long term
